### ISSUE TITLE
GMail OAuth flow does not work with non-localhost domains

### ISSUE THREAD

**Issue by largestprime** [mailpile/Mailpile#1861]
the gmail authentication pop up window works really well when using mailpile locally (on localhost port 33411) but I have mailpile running on a raspberry pi clone with pagekite running, and log into it from another computer. when using mailpile like this, the gmail pop up box works fine until the final step, when it returns the error "unable to connect to localhost 33411"

there is a workaround for this - to copy the URL from the pop up window, paste it in another window, then append "______ . pagekite.me" to the front, in the place of localhost, etc. but this is a tad inconvenient.

**Comment by BjarniRunar** [comment]
Thanks for reporting this. As remote access to a running Mailpile is a critical use-case, this is a 1.0 blocker.

**Comment by BjarniRunar** [comment]
Reopinating and renamifying.

**Comment by BjarniRunar** [comment]
It looks like we may be blocked by Google's requirement that every single URL that can be used for authentication, be pre-registered on their developers console. Maybe. We're still investigating.

**Comment by BjarniRunar** [comment]
After sleeping on it, I think we can work around this with a redirect service that redirects the redirects.

No joke. We'll just have to be very careful to make sure it doesn't become an abusable open redirect service.

**Comment by BjarniRunar** [comment]
Actually, that's not quite enough, the redirect endpoint receives the tokens.

However, we can arrange things in such a way that the service which receives the tokens doesn't know which account they apply to. Unless google have encoded that information into the tokens themselves, we can make the redirect completely blind to *whose* tokens it has, and the Mailpile client can then anonymously download them from there.

It's ugly and brittle and if Google are doing the wrong thing then we're still screwed... but it might work.

**Comment by BjarniRunar** [comment]
(this is all a problem because the Mailpile team does NOT want to have any user data or user credentials - the Google OAuth mechanism is designed for insecure "software as a service" setups where OAuth is used to grant some 3rd party access to all your data.)

**Comment by BjarniRunar** [comment]
Since it appears we cannot fix this in the near-term without both more research and a permanent hosted helper service of some sort, this is now too hard and out of scope for our 1.0 release candidate - we need more time.

I'm bumping this to 1.1 for now, will file a new 1.0 issue to improve the UI to inform users that 1.0 won't be able to support GMail and non-localhost Mailpile installations gracefully. :cry:

**Comment by BjarniRunar** [comment]
More options: The OAuth2 "OOB" method would make the login flow end with a page containing an auth code which the user could copy-paste into Mailpile. This would sort of work.

Search for "urn:ietf:wg:oauth:2.0:oob" here: https://developers.google.com/api-client-library/python/auth/installed-app

It'll be a shitty UX, but arguably better than GMail scaring people with admonishments of "legacy insecure apps" or and also arguably better than us running a man-in-the-middle authentication clearinghouse. This is probably our best option at this point, but I've reached out to friends at Google to see if anyone can dig up something better.
